### PR TITLE
Remove airstrike/paradrop beacon when the whole squad is shot down

### LINKED ISSUE #15740: Parabombs drop timer disappears when one plane shot down
When deploying the parabombs, a target icon with timer appears. It disappears after just one of the three planes is shot down. It should persist until all three planes are shot down. Otherwise, it seems as if the parabombs have failed completely.

An example where it's a pain: you may be waiting for the parabombs to tank the AA so you can swoop in with yaks/migs, but with no target+timer, you have to guess or track the planes.

### PR DESCRIPTION
**Pull request by dragunoff**
* Add a handler for `OnRemovedFromWorld` for the aircraft in the airstrike/paradrop. Inside the handler remove the beacon and camera only if the whole squad is dead.
* Add `RemoveCamera` and `RemoveBeacon` methods to the airstrike/paradrop support power traits.

Fixes #15740

### CODE REVIEW

**Comment by dragunoff** [comment]
I reworked this to separate the handling of `onRemovedFromWorld` and `onExitRange`.

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: Methods are `private` by default, so please drop the keyword.

File: OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs]: One level of indentation too much here.

File: OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs]: I wonder if this could be an issue when the beacon is removed between the `beacon == null` check and this frame end task running.

Comment (dragunoff) [reply OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs]: This is how it is now. I just copied the code. Should I move the `null` check inside the frame end task?

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs]: Fixed.

File: OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs]: Add another one. Not queuing a frame end task when `beacon` is `null` still makes sense, so that check doesn't have to go, imho.

Comment (dragunoff) [reply OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs]: Amended.

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: In practice it is guaranteed that the parent class's `Info` is an `AirstrikePowerInfo` to theres no need to cache this again here.  Better to cast it where you need it, which looks like only one extra place in this PR.

The better long term fix (for a different PR!) should be to redefine `SupportPower` as `public abstract class SupportPower<InfoType> : PausableConditionalTrait<InfoType> where InfoType : SupportPowerInfo` so that `Info` can be exposed directly as the correct type.

File: OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs]: Likewise here.

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: I don't understand why
```csharp
aircraftInRange[a] = false;
if (!aircraftInRange.Any(kv => kv.Value))
```
doesn't work here, which would simplify the fix to moving the `RemoveBeacon` call inside the `if` inside `onExitRange`.  Can you please expand this comment explaining why the check needs to be different, or re-merge the actions if it doesn't?

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: I copied this pattern from `IonCannonPower` because I needed a way to use the info in more than one method. I am not sure how to go about fixing that (I'm still new to C#).

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: Handling `onExitRange` and `onRemovedFromWolrd` in the same place leads to the bug that I set out to fix here. If an aircraft is shot down before the squad has reached target range that check will return `true` because all of the aircraft will be out of range. That's why checking for attack range is not relevant when aircraft actors are removed from world. That's why I'm checking if all aircraft `isDead`. 

Another way to go about it may be to remove actors from the dictionary when they killed and remove the camera and beacon once the dictionary is empty.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: Aha, now I see.

> Another way to go about it may be to remove actors from the dictionary when they killed and remove the camera and beacon once the dictionary is empty.

This sounds good to me

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs
Comment (pchote) [reply OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: I'll see if I can find time to fix this properly in the next couple of days so that you won't need to worry about it here.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: My idea for making `SupportPower` generic isn't compatible with the way it used in `SupportPowerManager`, so lets keep this part as-is.  We use this local-info pattern in several other support powers.

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: I have expanded the comment.

**Review by pchote — DISMISSED** [review]

**Comment by pchote** [comment]
LGTM but i'd like another pair of eyes on this before merging.

File: OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/SupportPowers/AirstrikePower.cs]: Minor style nit: I'd find `if (aircraftInRange.All(kv => kv.Key.IsDead))` easier to mentally parse than that double negation.

File: OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs]: Same here.

Comment (dragunoff) [reply OpenRA.Mods.Common/Traits/SupportPowers/ParatroopersPower.cs]: Amended.

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/fcae6e49ca503a7b2f2c979fa26a4d7c495c98f6...f89790e8c1b71847d0ec9570628c034aa4675876)

**Pull request merged**
